feat(outputs.opensearch): opensearch output plugin (#11958)
influxdata/telegraf: plugins/outputs/opensearch/README.md

  ##    drop    -- drop fields containing NaNs or infs
  ##    replace -- replace with the value in "float_replacement_value" (default: 0.0)
  ##               NaNs and inf will be replaced with the given number, -inf with the negative of that number
  # float_handling = "none"
  # float_replacement_value = 0.0

  ## Pipeline Config
  ## To use a ingest pipeline, set this to the name of the pipeline you want to use.
  # use_pipeline = "my_pipeline"

  ## Pipeline Name
  ## Additionally, you can specify a tag name using the notation (`{{.Tag "tag_name"}}`)
  ## If the tag does not exist, the default tag value will be used.
  # default_tag_value = ""
  ## If the tag does not exist, the default pipeline will be used as the
  ## pipeline. If no default pipeline is set, no pipeline is used for the
  ## metric.
  # default_pipeline = ""
```

### Required parameters

* `urls`: A list containing the full HTTP URL of one or more nodes from your
  OpenSearch instance.
* `index_name`: The target index for metrics. You can use the date format

For example: "telegraf-{{.Time.Format "2006-01-02"}}" would set it to telegraf-2023-07-27
You can also specify

> metric name (`{{Name}}`), tag value (`{{Tag "tag_name"}}`), field value (`{{Field "feild_name"}}`) 

Reviewer: ```suggestion
metric name (`{{ .Name }}`), tag value (`{{ .Tag "tag_name" }}`), field value (`{{ .Field "feild_name" }}`) 
```

  ## Setting to "0s" will disable the health check (not recommended in production)
  health_check_interval = "10s"
  ## Set the timeout for periodic health checks.
  # health_check_timeout = "1s"
  ## HTTP basic authentication details.
  ## HTTP basic authentication details
  # username = "telegraf"
  # password = "mypassword"
  ## HTTP bearer token authentication details
  # auth_bearer_token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9"

  ## Index Config
  ## The target index for metrics (Opensearch will create if it not exists).
  ## You can use the date specifiers below to create indexes per time frame.
  ## The metric timestamp will be used to decide the destination index name
  # %Y - year (2016)
  # %y - last two digits of year (00..99)
  # %m - month (01..12)
  # %d - day of month (e.g., 01)
  # %H - hour (00..23)
  # %V - week of the year (ISO week) (01..53)
  ## Additionally, you can specify a tag name using the notation {{tag_name}}
  ## which will be used as part of the index name. If the tag does not exist,
  ## the default tag value will be used.
  # index_name = "telegraf-{{host}}-%Y.%m.%d"
  # default_tag_value = "none"

>   index_name = "telegraf-%Y.%m.%d" # required.

Reviewer: I don't think inventing yet another templating language, which even mixes different notations will ease the use of it.

As other plugins already use the golang template mechanism, I would use this as well. Advantage would be to have more options for date specification.

See the [template processor](https://github.com/influxdata/telegraf/tree/master/plugins/processors/template) to use it's `TemplateMetric` And the example `index_name` formats will then look like this:
```toml
  index_name = 'telegraf-{{.Tag "host"}}-{{.Time.Format "2006-01-02"}}'
```

Author: Not a new invention, this is present in elasticsearch and supported already.

Reviewer: Can you show me the docs of ElasticSearch or OpenSearch where this format is explained?

Author: agree, will make the change

Reviewer: Thanks!

Author: Have made the changes, please take a look.

Reviewer 2: @mannukalra I fully agree with @Hipska, you really should reuse what is there as this will keep the code cleaner and will free you from extension requests to "your language".

Reviewer: Since @mannukalra started using golang templates, this conversation can be marked as resolved for me. (can't do it myself)

  ##    none    -- do not modify field-values (default); will produce an error
  ##               if NaNs or infs are encountered
  ##    drop    -- drop fields containing NaNs or infs
  ##    replace -- replace with the value in "float_replacement_value" (default: 0.0)
  ##               NaNs and inf will be replaced with the given number, -inf with the negative of that number
  # float_handling = "none"
  # float_replacement_value = 0.0

  ## Pipeline Config
  ## To use a ingest pipeline, set this to the name of the pipeline you want to use.
  # use_pipeline = "my_pipeline"

  ## Pipeline Name
  ## Additionally, you can specify a tag name using the notation {{tag_name}}
  ## which will be used as part of the pipeline name (e.g. "{{es_pipeline}}").
  ## If the tag does not exist, the default pipeline will be used as the
  ## pipeline. If no default pipeline is set, no pipeline is used for the
  ## metric.
  # default_pipeline = ""
```

### Required parameters

* `urls`: A list containing the full HTTP URL of one or more nodes from your
  OpenSearch instance.
* `index_name`: The target index for metrics. You can use the date specifiers
  below to create indexes per time frame.

> ```   %Y - year (2017)

Reviewer 2: ```suggestion
```text
   %Y - year (2017)
```

Author: done

  %y - last two digits of year (00..99)
  %m - month (01..12)
  %d - day of month (e.g., 01)
  %H - hour (00..23)
  %V - week of the year (ISO week) (01..53)
```

Additionally, you can specify dynamic index names by using tags with the
notation ```{{tag_name}}```. This will store the metrics with different tag
values in different indices. If the tag does not exist in a particular metric,

> the `default_tag_value` will be used instead.

Reviewer 2: Can you please provide some example for this? From reading it (and not knowing OpenSearch well) I cannot come up with something easily...

Author: Tags are used for grouping metrics, please refer https://www.elastic.co/guide/en/beats/metricbeat/current/add-tags.html

## Known issues

Integer values collected that are bigger than 2^63 and smaller than 1e21 (or in
this exact same window of their negative counterparts) are encoded by golang
JSON encoder in decimal format and that is not fully supported by OpenSearch
dynamic field mapping. This causes the metrics with such values to be dropped in
case a field mapping has not been created yet on the telegraf index. If that's
the case you will see an exception on OpenSearch side like this:

```json
{
  "error": {
    "root_cause": [
      {"type": "mapper_parsing_exception", "reason": "failed to parse"}
    ],
    "type": "mapper_parsing_exception",
    "reason": "failed to parse",
    "caused_by": {
      "type": "illegal_state_exception",
      "reason": "No matching token for number_type [BIG_INTEGER]"
    }
  },
  "status": 400
}
```

The correct field mapping will be created on the telegraf index as soon as a
supported JSON value is received by OpenSearch, and subsequent insertions
will work because the field mapping will already exist.

This issue is caused by the way OpenSearch tries to detect integer fields,
and by how golang encodes numbers in JSON. There is no clear workaround for this

> at the moment.

Reviewer 2: For my understanding: Are you saying that if you have numbers `> int64` the JSON encoder will output a number in exponential form? How do you represent those integers anyway?

Author: More than representation it is a limitation of elastic/opensearch wrt. Max/Min long value. As you know this is copied from elasticseach, please let know if you feel content needs to modified.

# OpenSearch Output Plugin

This plugin writes to [OpenSearch](https://opensearch.org/) via HTTP

It supports OpenSearch releases from 1 and 2. Future comparability with 1.x is
not guaranteed and instead will focus on 2.x support. Consider using the
existing Elasticsearch plugin for 1.x.

## Global configuration options <!-- @/docs/includes/plugin_config.md -->

In addition to the plugin-specific configuration settings, plugins support
additional global and plugin configuration settings. These settings are used to
modify metrics, tags, and field or create aliases and configure ordering, etc.
See the [CONFIGURATION.md][CONFIGURATION.md] for more details.

[CONFIGURATION.md]: ../../../docs/CONFIGURATION.md#plugins

## Configuration

```toml @sample.conf
# Configuration for OpenSearch to send metrics to.
[[outputs.OpenSearch]]
  ## URLs
  ## The full HTTP endpoint URL for your OpenSearch instance. Multiple URLs can
  ## be specified as part of the same cluster, but only one URLs is used to
  ## write during each interval.
  urls = ["http://node1.os.example.com:9200"]

>   ## Index Name
>   ## The target index for metrics (OpenSearch will create if it not exists).
>   ## You can use the date specifiers below to create indexes per time frame.
>   ## The metric timestamp will be used to decide the destination index name
>   ##   %Y - year (2016)
>   ##   %y - last two digits of year (00..99)
>   ##   %m - month (01..12)
>   ##   %d - day of month (e.g., 01)
>   ##   %H - hour (00..23)
>   ##   %V - week of the year (ISO week) (01..53)
>   ## For example: telegraf-%Y.%m.%d would set it to telegraf-2006-01-02
>   index_name = ""
> 

Reviewer 2: The idea of @Hipska is to reuse go templates so this should be
```suggestion
  ## Target index name for metrics (OpenSearch will create if it not exists).
  ## This is a Golang template (see https://pkg.go.dev/text/template) to
  ## access the metric name (`{{.Name}}`), a tag value (`{{.Tag "name"}}`) ,
  ## a field value (`{{.Field "name"}}`) or the timestamp (`{{.Time}}`).
  ## For example: "telegraf-{{.Time.Format "2006-01-02"}}" would set it to telegraf-2023-07-27
  index_name = ""

```

Author: Used with little change as {{Name}}`), (`{{Tag "name"}}`), (`{{Field "name"}}`) are not prefixed with a .
because of the way string-template's implementation wrt. function.

Reviewer: If using `templateMetric`, then these methods come available to the object. This is a crucial part.

Author: made the change, working as expected.

# Opensearch Output Plugin

This plugin writes to [Opensearch](https://opensearch.org/) via HTTP using
[Elastic](http://olivere.github.io/elastic/) client API.

> It supports Opensearch releases from 1.x up to 2.x.

Reviewer 3: Was there a 0.x release? If not we can just say:

```suggestion
It supports OpenSearch releases 1 and 2. Future comparability with 1.x is not
guaranteed and instead will focus on 2.x support. Consider using the existing
Elasticsearch plugin for 1.x.
```

Author: done

  ## This option can have the following values:
  ##    none    -- do not modify field-values (default); will produce an error if NaNs or infs are encountered
  ##    drop    -- drop fields containing NaNs or infs
  ##    replace -- replace with the value in "float_replacement_value" (default: 0.0)
  ##               NaNs and inf will be replaced with the given number, -inf with the negative of that number
  # float_handling = "none"
  # float_replacement_value = 0.0

  ## Pipeline Config
  ## To use a ingest pipeline, set this to the name of the pipeline you want to use.
  # use_pipeline = "my_pipeline"
  ## Additionally, you can specify a tag name using the notation {{tag_name}}
  ## which will be used as part of the pipeline name. If the tag does not exist,
  ## the default pipeline will be used as the pipeline. If no default pipeline is set,
  ## no pipeline is used for the metric.
  # use_pipeline = "{{es_pipeline}}"
  # default_pipeline = "my_pipeline"
```

### Permissions

If you are using authentication within your Opensearch cluster, you need to
create a account and create a role with at least the manage role in the Cluster

> Privileges category.  Overwise, your account will not be able to connect to your

Author: corrected

Opensearch cluster and send logs to your cluster.  After that, you need to

> add "create_indice" and "write" permission to your specific index pattern.

Author: corrected

* `overwrite_template`: Set to true if you want telegraf to overwrite an
  existing template.
* `force_document_id`: Set to true will compute a unique hash from as
  sha256(concat(timestamp,measurement,series-hash)),enables resend or update
  data withoud ES duplicated documents.
* `float_handling`: Specifies how to handle `NaN` and infinite field
  values. `"none"` (default) will do nothing, `"drop"` will drop the field and
  `replace` will replace the field value by the number in
  `float_replacement_value`
* `float_replacement_value`: Value (defaulting to `0.0`) to replace `NaN`s and
  `inf`s if `float_handling` is set to `replace`. Negative `inf` will be
  replaced by the negative value in this number to respect the sign of the
  field's original value.
* `use_pipeline`: If set, the set value will be used as the pipeline to call
  when sending events to opensearch. Additionally, you can specify dynamic
  pipeline names by using tags with the notation ```{{tag_name}}```.  If the tag
  does not exist in a particular metric, the `default_pipeline` will be used
  instead.
* `default_pipeline`: If dynamic pipeline names the tag does not exist in a

>   particular metric, this value will be used instead.

Author: Actually this is followed the same way in elasticsearch, please highlight the section if you still feel we should remove some parts.

  urls = ["http://node1.os.example.com:9200"]

  ## Index Name
  ## Target index name for metrics (OpenSearch will create if it not exists).
  ## This is a Golang template (see https://pkg.go.dev/text/template)
  ## You can also specify

>   ## metric name (`{{Name}}`), tag value (`{{Tag "tag_name"}}`), field value (`{{Field "feild_name"}}`) 

Reviewer: I mentioned multiple times earlier, you need to provide a `templateMetric` instance to the `Exec` method..

>   ## metric name (`{{Name}}`), tag value (`{{Tag "tag_name"}}`), field value (`{{Field "feild_name"}}`) 

Author: I am also fed up of explaining, give we some appropriate reference.

>   ## metric name (`{{Name}}`), tag value (`{{Tag "tag_name"}}`), field value (`{{Field "feild_name"}}`) 

Author: finally made the expected way.
